downtimes.1
.\"-
.\" Copyright (c) 2009-2016 Janne Snabb. All rights reserved.
.\"
.\" This software is licensed under the terms and conditions of the
.\" Simplified BSD License. You should have received a copy of that
.\" license along with this software.
.\"
.TH DOWNTIMES 1 "2016-05-24" "version 1.0"
.SH NAME
downtimes \- display system downtime records made by
.BR downtimed (8)
.SH SYNOPSIS
.B downtimes
.RB [\| \-d
.IR downtimedbfile \|]
.RB [\| \-f
.IR timefmt \|]
.RB [\| \-n
.IR num \|]
.RB [\| \-s
.IR sleep \|]
.RB [\| \-u \|]
.br
.B downtimes
.B \-v
.br
.B downtime
.RB [\| \-d
.IR downtimedbfile \|]
.RB [\| \-f
.IR timefmt \|]
.RB [\| \-n
.IR num \|]
.RB [\| \-s
.IR sleep \|]
.RB [\| \-u \|]
.br
.B downtime
.B \-v
.SH DESCRIPTION
The
.B downtimes
command displays the contents of the downtime database in human readable
format. The downtime database is maintained by
.BR downtimed (8).
.PP
The singular form
.B downtime
displays details only about the last time the system was down,
whereas the plural form
.B downtimes
displays all records in the database.
The
.B \-n
option can be used to specify the number of downtime
records to display.
.SH OPTIONS
.TP
.B \-d \fIdowntimedbfile\fR
Use the specified downtime database file instead of the system default.
.TP
.B \-f \fItimefmt\fR
Specify the time and date format to use when reporting using
.BR strftime (3)
syntax. The default is "%F %T".
.TP
.B \-n \fInum\fR
Define how many latest downtime records to output. Default is all.
.TP
.B \-s \fIsleep\fR
Calculate the approximate crash time by specifying what was the
sleep value of
.BR downtimed (8).
.TP
.B \-u
Display times in UTC.
.TP
.B \-v
Display the program version number, copyright message and the default
settings.
.SH EXIT STATUS
The program exits 0 on success, and >0 if an error occurs.
.SH SEE ALSO
.BR downtimed (8),
.BR uptime (1),
.ie !d pdfhref \
https://dist.epipe.com/downtimed/
.el \
.pdfhref W https://dist.epipe.com/downtimed/
.SH BUGS
The reporting accuracy in case of a system crash depends on how often the
time stamp is updated.
.SH COPYRIGHT
Copyright \(co 2009\-2016 Janne Snabb. All rights reserved.
.PP
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
.PP
1. Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.
.PP
2. Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution.
.PP
THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
SUCH DAMAGE.
.\" eof


NEWS
               downtimed - system downtime reporting tool

                           version history

version 1.0 released 2016-05-24

 - This is mainly a maintenance release. There are no significant new
   features. The software appears to be so stable that it is the time to
   release 1.0 now.

 - Enable downtime(1) command and manual page. It was previously commented
   out but enabled in Debian and Ubuntu packaging. The singular form
   displays just the most recent downtime entry. The original idea was that
   this is the counterpart of the uptime(1) command.

 - Add downtimed(8) command line option -F for enabling seamless
   operation with modern system service supervisors such as systemd(8) and
   others. This option disables calling daemon(3) and thus fork(2)ing into
   the background. In systemd(8) terminology this is a [Service] Type=
   "simple".

 - Eliminate #ifndef __APPLE__ in the code. OS X users must now add the -F
   option described above to their downtimed(8) launchd(8) configuration.
   The example configuration startup-scripts/com.epipe.downtimed.plist was
   updated accordingly.

 - Add systemd.service(5) file in startup-scripts/downtimed.service.

 - Add -f command line option to usage help strings. Reported and a patch
   supplied by Mats Erik Andersson. Thanks!

 - Allow disabling pid file creation with -p none.

 - Update autotools produced scripts, build-aux.

 - Moved code hosting from Launchpad to GitHub. The new location is:

     https://github.com/snabb/downtimed

   Also README was changed to README.md.

 - Miscellaneous minor improvements and changes in the documentation.


version 0.6 released 2013-07-27

 - Added startup scripts for Debian, Red Hat (RHEL/Fedora/Scientific
   Linux/CentOS/etc.), Arch Linux and OpenIndiana/OpenSolaris.

 - Implemented -f command line option for downtimes(1) and downtimed(8).
   It can be used to specify the desired time and date format for
   reporting. The default remains "%F %T".

 - Implemented -u command line option for downtimes(1) for reporting
   times in UTC. Analogous to "date -u".

 - GNU/Hurd compatibility fixes contributed by Mats Erik Andersson.

 - FreeBSD 9 compatibility fix by Douglas Thrift.


version 0.5 released 2011-03-02

 - Fixed a bug which prevented creation of the required time stamp
   files in case command line option -S was set.

 - Determine boot time on some SVR4 systems such as Solaris.

 - Compatibility for systems which do not have flock() but have lockf(),
   such as Solaris.

 - Compatibility for systems without facilitynames in <syslog.h>, such
   as Solaris.

 - Compatibility daemon(3) function for OSes which do not have it, such
   as Solaris.

 - Compatibility for systems which lack futimes(3) function, such as
   Solaris.

 - Log an error message if system boot time can not be determined.

 - Remove dependency on be64toh(3) and betoh64(3). Or betoh64(3) and
   htobe64(3). Or OSSwapBigToHostInt64(3) and OSSwapHostToBigInt64(3).
   The code became a mess because 64 bit byte swapping is not standardized.

   Also some commonly used GNU/Linux distributions such as RHEL/CentOS 5.5
   have such an old glibc that it does not include any of these functions.

   Now we just implement our own byte swapping macro and ignore whatever
   is or is not available from the system. This improves portability
   on many platforms, for example on Red Hat (RHEL) and Solaris.

 - Improve README documentation regarding startup scripts.

 - Example startup scripts have been moved to the "startup-scripts"
   sub-directory.

 - openSUSE startup script by Federico Lucifredi <flucifredi@acm.org>.

 - OpenBSD compatibility improvements.

 - Debian GNU/kFreeBSD compatibility improvements.

 - Use volatile sig_atomic_t if available for flags set by signal handlers.

 - Other minor fixes.


version 0.4 released 2010-12-25

 - Do not create the data directory /var/{db,lib}/downtimed in the
   daemon. It should be created by the system administrator or the
   package maintainer instead. Problem reported by Mats Erik Andersson
   <openbsd@gisladisker.se>.

 - OpenBSD support and other minor fixes by Mats Erik
   Andersson <openbsd@gisladisker.se>.

 - Debian GNU/kFreeBSD support by Mats Erik Andersson
   <openbsd@gisladisker.se>.

 - Change backwards lseek(2) to forward lseek(2) as the backwards seek
   fails on some platforms, by Mats Erik Andersson <openbsd@gisladisker.se>.

 - Other minor fixes.


version 0.3 released 2010-10-26

 - Mac OS X support by Henrik Ahlgren <pablo@seestieto.com>

 - Fix man page rendering on various OSes such as FreeBSD (we were
   using non-standard .SY, .YS and .OP macros) [this has no
   effect on GNU/Linux, no need to upgrade]

 - Fix printf format string to avoid compiler warnings


version 0.2 released 2010-09-22

 - implemented downtimes(1) command for displaying the contents of
   the downtime database in human readable format

 - implemented downtime database in the downtimed(8) daemon

 - moved the location of data files from /var/lib/misc/downtimed/
   to /var/lib/downtimed/ on Linux.

 - the default sleep time between time stamp updates is now 15 seconds
   instead of 5 as previously

 - minor documentation improvements


version 0.1 released 2010-05-22

 - initial release



downtimes.c
/*-
 * Copyright (c) 2009-2016 Janne Snabb. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *
 * Software web site:
 *   https://dist.epipe.com/downtimed/
 *
 */

/* Include config.h in case we use autoconf. */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

/* Standard includes that we need */

#include <sys/file.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <inttypes.h>

#ifdef HAVE_PATHS_H
#include <paths.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sysexits.h>
#include <time.h>
#include <unistd.h>

#include "downtimedb.h"

/* Some global defines */

#define	PROGNAME "downtimes"

/* PACKAGE_VERSION is defined by autoconf, if used. */
#ifdef PACKAGE_VERSION
#define	PROGVERSION PACKAGE_VERSION
#else
#define	PROGVERSION "0.0undef"
#endif

/* Function prototypes */

int		main(int, char *[]);
static void	report(int64_t, int, int64_t);
static void	version(void);
static void	usage(void);
static void	parseargs(int, char *[]);

/* Command line arguments with their defaults */

static long	cf_sleep = 0; /* adjust crash time according to sleep value */
static char *	cf_downtimedbfile = PATH_DOWNTIMEDBFILE;
static long	cf_n = -1;         /* number of downtime records to display */
static char *	cf_timefmt = FMT_DATETIME;
static int	cf_utc = 0;                  /* set to display times in UTC */

/*
 * downtimes: display system downtime records made by downtimed(8)
 */

int
main(int argc, char *argv[])
{
	struct downtimedb dbent;
	struct stat sb;
	int64_t tdown, tadjust;
	int fd, ret, crashed;

	/* parse command line arguments */
	parseargs(argc, argv);

	if ((fd = open(cf_downtimedbfile, O_RDONLY)) < 0) {
		fputs("Maybe the system has not been down yet?\n", stderr);
		err(EX_NOINPUT, "can not open %s", cf_downtimedbfile);
	}

	if (fstat(fd, &sb) < 0)
		err(EX_NOINPUT, "can not stat %s", cf_downtimedbfile);

	if (sb.st_size % sizeof(struct downtimedb) != 0)
		errx(EX_DATAERR, "%s is corrupted", cf_downtimedbfile);

	if ((cf_n == -1)
	    || (cf_n > sb.st_size / sizeof(struct downtimedb) / 2))
		cf_n = sb.st_size / sizeof(struct downtimedb) / 2;

	if (lseek(fd, sb.st_size - (cf_n * sizeof(struct downtimedb) * 2),
	    SEEK_SET) < 0)
		err(EX_DATAERR, "can not seek %s", cf_downtimedbfile);

	tdown = 0;
	tadjust = cf_sleep / 2;

	while ((ret = downtimedb_read(fd, &dbent)) > 0) {
		switch (dbent.what) {
		case DOWNTIMEDB_WHAT_SHUTDOWN:
			if (tdown != 0)
				report(dbent.when, 0, 0);
			tdown = dbent.when;
			crashed = 0;
			break;
		case DOWNTIMEDB_WHAT_CRASH:
			if (tdown != 0)
				report(dbent.when + tadjust, 1, 0);
			tdown = dbent.when;
			crashed = 1;
			break;
		case DOWNTIMEDB_WHAT_UP:
			report(crashed ? tdown + tadjust : tdown,
			    crashed, dbent.when);
			tdown = 0;
			break;
		case DOWNTIMEDB_WHAT_NONE:
		default:
			break;
		}
	}

	if (tdown != 0)
		report(tdown, crashed, 0);

	if (ret < 0)
		err(EX_DATAERR, "error reading %s", cf_downtimedbfile);

	close(fd);
	exit(EX_OK);
}

/* Output one line of downtime report */

static void
report(int64_t td, int crashed, int64_t tu)
{
	printf("%s %s -> ", crashed ? "crash" : "down ", 
	    timestr_abs((time_t) td, cf_timefmt, cf_utc));

	/* Note that the printf() above and below is intentionally split
	   into two parts because timestr_abs() clobbers the returned
	   static string on each call. */

	printf("up %s ", timestr_abs((time_t) tu, cf_timefmt, cf_utc));

	/* timestr_int() returns string representing a relative time (time
	   period) such as 21+06:11:38 or 06:11:38 */

	if (tu != 0 && td != 0)
		printf("= %11s (%"PRIu64" s)\n",
		    timestr_int((time_t)(tu - td)), tu - td);
	else
		printf("= %11s (? s)\n", "unknown");
}

/* Usage help & exit */

static void
usage()
{

	fputs("usage: " PROGNAME " [-v] [-d downtimedbfile] [-f timefmt] "
	    "[-n num] [-s sleep] [-u]\n", stderr);
	exit(EX_USAGE);
}

/* Output version information, default settings & exit */

static void
version()
{

	puts(PROGNAME " " PROGVERSION " - display system downtime records "
	    "made by downtimed(8)\n");

	puts("Copyright (c) 2009-2016 Janne Snabb. "
	    "All rights reserved.");

	puts("This software is licensed under the terms and conditions of the");
	puts("Simplified BSD License. You should have received a copy of that");
	puts("license along with this software.\n");

	puts("Default settings:");
	printf("  downtimedbfile = %s\n", cf_downtimedbfile);
	printf("  num = %ld\n", cf_n);
	printf("  sleep = %ld\n", cf_sleep);
	printf("  timefmt = %s\n", cf_timefmt);
	printf("  utc = %d\n", cf_utc);

#ifdef PACKAGE_URL
	puts("\nSee the following web site for more information and updates:");
	puts("  " PACKAGE_URL "\n");
#endif
	exit(EX_OK);
}

/* Handle command line arguments */

static void
parseargs(int argc, char *argv[])
{
	int c;
	char *p;

	if (strlen(argv[0]) > 0 && argv[0][strlen(argv[0])-1] != 's')
		cf_n = 1;

	while ((c = getopt(argc, argv, "d:f:n:s:uvh?")) != -1) {
		switch (c) {
		case 'd':
			cf_downtimedbfile = optarg;
			break;
		case 'f':
			cf_timefmt = optarg;
			break;
		case 'n':
			p = NULL;
			errno = 0;
			cf_n = strtol(optarg, &p, 10);
			if ((p != NULL && *p != '\0') || errno != 0)
				errx(EX_USAGE, "-n argument is not a number");
			break;
		case 's':
			p = NULL;
			errno = 0;
			cf_sleep = strtol(optarg, &p, 10);
			if ((p != NULL && *p != '\0') || errno != 0)
				errx(EX_USAGE, "-s argument is not a number");
			break;
		case 'u':
			cf_utc = 1;
			break;
		case 'v':
			version();
			/* NOTREACHED */
			break;
		case 'h':
		case '?':
		default:
			usage();
			/* NOTREACHED */
			break;
		}
	}
	if (argc != optind)
		usage();
}

/* eof */
